py-polars/polars/io/delta.py

```diff
@@ -313,3 +316,109 @@ def _check_if_delta_available() -> None:
         raise ImportError(
             "deltalake is not installed. Please run `pip install deltalake>=0.9.0`."
         )
+
+
+def _reconstruct_field_type(
+    field: pa.Field,
+    field_head: pa.Field,
+    reconstructed_field: list | None = None,
+) -> pa.Field:
+    """
+    Recursive function that traverses through pyArrow fields.
+
+    Parameters
+    ----------
+    field : pa.Field
+        field to analyze
+    field_head : pa.Field
+        field head to retain column names
+    reconstructed_field : List, optional
+        Collection of dtypes to reconstruct a field, by default None
+
+    Returns
+    -------
+    pa.Field
+        Reconstructed field
+    """
+    if isinstance(field.type, pa.TimestampType):
+        if reconstructed_field is None:
+            return pa.field(
+                name=field.name,
+                type=pa.timestamp(
+                    "us"
+                    #   , tz=field.type.tz
+                ),
+            )
+        else:
+            reconstructed_field.append(
+                pa.timestamp(
+                    "us",
+                    # tz=field.type.tz
+                )
+            )
+            return pa.field(
+                name=field_head.name,
+                type=reduce(lambda x, y: y(x), reversed(reconstructed_field)),
+            )
+    elif "uint" in str(field.type) and not any(
+        isinstance(field.type, dtype) for dtype in [pa.StructType, pa.LargeListType]
+    ):
+        if reconstructed_field is None:
+            return pa.field(
+                name=field.name, type=getattr(pa, str(field.type).strip("u"))()
+            )
+        else:
+            reconstructed_field.append(getattr(pa, str(field.type).strip("u"))())
+            return pa.field(
+                name=field_head.name,
+                type=reduce(lambda x, y: y(x), reversed(reconstructed_field)),
+            )
```

Comment:
I don't like operating on strings... just define a mapping between types and use that, with some recursive logic to iterate through nested types.

Reply:
While working on something related, maybe we should not do Uint -> Int, since it's possible that a number in Uint is to big to cast into Int. Before I had in my mind that the conversion would be safe, but it's not in some cases.

Reply:
We are not casting, we are just determining what the matching schema would be if it were written to a delta table.

If the value does not fit, the `deltalake` package (which will try to cast it) will throw an error.

Reply:
> We are not casting, we are just determining what the matching schema would be if it were written to a delta table.
> 
> If the value does not fit, the `deltalake` package (which will try to cast it) will throw an error.

Ah you're right. I thought Delta-RS would just silently still write the table.

Reply:
@stinodego regarding your comment. An int, uint, string, largestring are all instance of pa.DataType, from what I can see the only way to distinguish which is which is to check the name of pa.DataType using str(datatype), or datatype.__str__(), or datatype.equals(pa.Uint32)